File: Nori/Store/SensitiveVault.swift
import Foundation
import Observation

/// Secrets live here, never in SwiftData: masked in the list, pasteable, gone after ten minutes.
@MainActor
@Observable
final class SensitiveVault {
    struct Entry: Sendable {
        let id: UUID
        let match: SecretDetector.Match
        let mask: String
        let draft: ClipDraft
        let capturedAt: Date
        var expiresAt: Date
    }

    static let lifetime: TimeInterval = 10 * 60

    private(set) var entries: [Entry] = []
    private(set) var version = 0
    private var sweepTimer: Timer?

    init() {
        let timer = Timer(timeInterval: 30, repeats: true) { [weak self] _ in
            MainActor.assumeIsolated { self?.sweep() }
        }
        RunLoop.main.add(timer, forMode: .common)
        sweepTimer = timer
    }

    var rows: [ClipRow] {
        entries.map { entry in
            ClipRow(
                id: entry.id, source: .sensitive, kind: entry.draft.kind,
                title: entry.mask, searchText: "",
                sourceBundleID: entry.draft.sourceBundleID, sourceAppName: entry.draft.sourceAppName,
                isFromUniversalClipboard: entry.draft.isFromUniversalClipboard,
                firstCopiedAt: entry.capturedAt, lastCopiedAt: entry.capturedAt, copyCount: 1, pinnedAt: nil,
                byteCount: entry.draft.byteCount, characterCount: entry.draft.characterCount,
                lineCount: entry.draft.lineCount, isRichText: false, isTruncated: false,
                linkURL: nil, colorHex: nil, fileURLs: [], imagePixelSize: nil, thumbnail: nil,
                expiresAt: entry.expiresAt, ghostReason: nil
            )
        }
    }

    /// Add a secret, or extend the life of an identical one.
    @discardableResult
    func add(_ sensitive: SensitiveDraft, now: Date = .now) -> UUID {
        if let index = entries.firstIndex(where: { $0.draft.contentHash == sensitive.draft.contentHash }) {
            entries[index].expiresAt = now.addingTimeInterval(Self.lifetime)
            let entry = entries.remove(at: index)
            entries.insert(entry, at: 0)
            version += 1
            return entry.id
        }
        let entry = Entry(id: UUID(), match: sensitive.match, mask: sensitive.mask, draft: sensitive.draft,
                          capturedAt: now, expiresAt: now.addingTimeInterval(Self.lifetime))
        entries.insert(entry, at: 0)
        version += 1
        return entry.id
    }

    func entry(id: UUID) -> Entry? {
        entries.first { $0.id == id }
    }

    /// Pasting a secret keeps it around for another ten minutes.
    func touch(id: UUID, now: Date = .now) {
        guard let index = entries.firstIndex(where: { $0.id == id }) else { return }
        entries[index].expiresAt = now.addingTimeInterval(Self.lifetime)
        let entry = entries.remove(at: index)
        entries.insert(entry, at: 0)
        version += 1
    }

    func remove(id: UUID) {
        entries.removeAll { $0.id == id }
        version += 1
    }

    func removeAll() {
        entries.removeAll()
        version += 1
    }

    func sweep(now: Date = .now) {
        let before = entries.count
        entries.removeAll { $0.expiresAt <= now }
        if entries.count != before { version += 1 }
    }
}

File: windows/Nori.Core/SensitiveVault.cs
namespace Nori.Core;

/// <summary>
/// Secrets live here, never in the database: masked in the list, pasteable, gone after ten minutes.
/// The app runs <see cref="Sweep"/> on a timer; Core keeps the logic clock-driven so it is testable.
/// </summary>
public sealed class SensitiveVault
{
    public sealed class Entry
    {
        public required Guid Id { get; init; }
        public required SecretDetector.Match Match { get; init; }
        public required string Mask { get; init; }
        public required ClipDraft Draft { get; init; }
        public required DateTimeOffset CapturedAt { get; init; }
        public DateTimeOffset ExpiresAt { get; set; }
    }

    public static readonly TimeSpan Lifetime = TimeSpan.FromMinutes(10);

    private readonly List<Entry> _entries = [];

    public IReadOnlyList<Entry> Entries => _entries;

    /// <summary>Bumped on every visible change so observers can refresh cheaply.</summary>
    public int Version { get; private set; }

    public IReadOnlyList<ClipRow> Rows => _entries.Select(entry => new ClipRow
    {
        Id = entry.Id,
        Source = RowSource.Sensitive,
        Kind = entry.Draft.Kind,
        Title = entry.Mask,
        SearchText = string.Empty,
        SourceApp = entry.Draft.SourceApp,
        SourceAppName = entry.Draft.SourceAppName,
        SourceAppPath = entry.Draft.SourceAppPath,
        FirstCopiedAt = entry.CapturedAt,
        LastCopiedAt = entry.CapturedAt,
        CopyCount = 1,
        PinnedAt = null,
        ByteCount = entry.Draft.ByteCount,
        CharacterCount = entry.Draft.CharacterCount,
        LineCount = entry.Draft.LineCount,
        ExpiresAt = entry.ExpiresAt,
    }).ToList();

    /// <summary>Add a secret, or extend the life of an identical one.</summary>
    public Guid Add(SensitiveDraft sensitive, DateTimeOffset now)
    {
        var index = _entries.FindIndex(e => e.Draft.ContentHash == sensitive.Draft.ContentHash);
        if (index >= 0)
        {
            var existing = _entries[index];
            existing.ExpiresAt = now + Lifetime;
            _entries.RemoveAt(index);
            _entries.Insert(0, existing);
            Version++;
            return existing.Id;
        }
        var entry = new Entry
        {
            Id = Guid.NewGuid(),
            Match = sensitive.Match,
            Mask = sensitive.Mask,
            Draft = sensitive.Draft,
            CapturedAt = now,
            ExpiresAt = now + Lifetime,
        };
        _entries.Insert(0, entry);
        Version++;
        return entry.Id;
    }

    public Entry? Find(Guid id) => _entries.FirstOrDefault(e => e.Id == id);

    /// <summary>Pasting a secret keeps it around for another ten minutes.</summary>
    public void Touch(Guid id, DateTimeOffset now)
    {
        var index = _entries.FindIndex(e => e.Id == id);
        if (index < 0) return;
        var entry = _entries[index];
        entry.ExpiresAt = now + Lifetime;
        _entries.RemoveAt(index);
        _entries.Insert(0, entry);
        Version++;
    }

    public void Remove(Guid id)
    {
        _entries.RemoveAll(e => e.Id == id);
        Version++;
    }

    public void RemoveAll()
    {
        _entries.Clear();
        Version++;
    }

    /// <summary>Drops expired entries; expiry is inclusive (<c>ExpiresAt &lt;= now</c>).</summary>
    public void Sweep(DateTimeOffset now)
    {
        var removed = _entries.RemoveAll(e => e.ExpiresAt <= now);
        if (removed > 0) Version++;
    }
}
